### PR TITLE
Add qchem basis rotation function

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR adds a function for basis rotation grouping of a molecular Hamiltonian.

**Description of the Change:**
The function `basis_rotation` is added to qchem.factorization module.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qchem/factorization.py [hunk 1]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+
+        The orbital basis can be rotated such that each :math:`T` and :math:`L^{(r)}` matrix is
+        diagonal. The Hamiltonian can then be written following Eq. (2) of
+        [`npj Quantum Information, 7, 23 (2021) <https://www.nature.com/articles/s41534-020-00341-7>`_]
+        as
+
+        .. math::
+
+            H = U_0 \left ( \sum_p d_p n_p \right ) U_0^{\dagger} + \sum_r^R U_r \left ( \sum_{pq}
+            d_{pq}^{(r)} n_p n_q \right ) U_r^{\dagger}
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:252]:
```suggestion
            d_{pq}^{(r)} n_p n_q \right ) U_r^{\dagger},
```

File: pennylane/qchem/factorization.py [hunk 2]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+            n_p = \frac{1-Z_p}{2}
+
+        where :math:`Z_p` is the Pauli :math:`Z` operator applied to qubit :math:`p`. This gives
+        the qubit Hamiltonian
+
+        .. math::
+
+           H = U_0 \left ( \sum_p O_p^{(0)} \right ) U_0^{\dagger} + \sum_r^R U_r \left ( \sum_{q} O_q^{(r)} \right ) U_r^{\dagger},
+
+        where :math:`O = \sum_i c_i P_i` is a linear combination of Pauli words :math:`P_` that are
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:269]:
```suggestion
        where :math:`O = \sum_i c_i P_i` is a linear combination of Pauli words :math:`P_i` that are
```

File: pennylane/qchem/factorization.py [hunk 3]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+        .. math::
+
+            n_p = \frac{1-Z_p}{2}
+
+        where :math:`Z_p` is the Pauli :math:`Z` operator applied to qubit :math:`p`. This gives
+        the qubit Hamiltonian
+
+        .. math::
+
+           H = U_0 \left ( \sum_p O_p^{(0)} \right ) U_0^{\dagger} + \sum_r^R U_r \left ( \sum_{q} O_q^{(r)} \right ) U_r^{\dagger},
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:267]: What is meant by `p^{(0)}` and `q^{(r)}`?

File: pennylane/qchem/factorization.py [hunk 4]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+    """
+    two_electron = np.swapaxes(two_electron, 1, 3)
+
+    _, eigvals_m, eigvecs_m = qml.qchem.factorize(two_electron, tol_factor, 0.0)
+
+    t_matrix = one_electron - 0.5 * np.einsum("illj", two_electron)
+    t_eigvals, t_eigvecs = np.linalg.eigh(t_matrix)
+
+    eigvals = [np.array(t_eigvals)] + [np.outer(x, x).flatten() * 0.5 for x in eigvals_m]
+    eigvecs = [t_eigvecs] + eigvecs_m
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:284]: What doe this intend to do? `t_eigvecs` will be an array where its columns are the eigenvectors, but what about `eigvecs_m`?

File: pennylane/qchem/factorization.py [hunk 5]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+                if i == j:
+                    ops_l_ += cc * (
+                        qml.Identity(i) - qml.PauliZ(i) - qml.PauliZ(j) + qml.Identity(i)
+                    )
+                else:
+                    ops_l_ += cc * (
+                        qml.Identity(i)
+                        - qml.PauliZ(i)
+                        - qml.PauliZ(j)
+                        + qml.grouping.pauli_mult_with_phase(qml.PauliZ(i), qml.PauliZ(j))[0]
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:305]: Wouldn't putting `i=j` here give Identity? If yes, then why are we branching inside the loop?

Comment (obliviateandsurrender) [reply pennylane/qchem/factorization.py:305]: Also, we should be able to reduce the iterations here using `(i, j) = (j, i)`.

File: pennylane/qchem/factorization.py [hunk 6]
```diff
@@ -169,3 +170,145 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+    rotation unitaries obtained with the basis rotation grouping method.
+
+    Args:
+        one_electron (array[float]): one-electron integral matrix in the molecular orbital basis
+        two_electron (array[array[float]]): two-electron integral tensor in the molecular orbital
+            basis arranged in chemist notation
+        tol_factor (float): threshold error value for discarding the negligible factors
+
+    Returns:
+        tuple(list[array[float]], list[list[Observable]], list[array[float]]): tuple containing the
```
Comment (obliviateandsurrender) [inline pennylane/qchem/factorization.py:186]: How does the last return args: `list[array(float)]` represent unitaries?

File: tests/qchem/test_factorization.py [hunk 7]
```diff
@@ -167,3 +167,37 @@ def test_empty_error(two_tensor):
 
     with pytest.raises(ValueError, match="All eigenvectors are discarded."):
         qml.qchem.factorize(two_tensor, 1e-5, 1e1)
+
+
+@pytest.mark.parametrize(
+    ("one_matrix", "two_tensor", "tol_factor", "coeffs_ref"),
+    [
```
Comment (obliviateandsurrender) [inline tests/qchem/test_factorization.py:174]: Are these obtained manually or via `openfermion`?

File: pennylane/qchem/factorization.py (hunk 3 above)
Comment (soranjh) [reply pennylane/qchem/factorization.py:267]: These are `O_p^{(0)}` and `O_q^{(r)}`, where O is defined in the line below the equation as `O = \sum_i c_i P_i`. The labels `0` and `r` are added for consistency with the unitaries `U_0` and `U_r`. Since `r` starts from `1`, we have used `0` for the first unitary, following the original paper.

File: pennylane/qchem/factorization.py (hunk 4 above)
Comment (soranjh) [reply pennylane/qchem/factorization.py:284]: eigvecs is a list containing all the eigenvectors: those from the T matrix, which specify the one-body terms and those from factorising the two-body integrals.

File: pennylane/qchem/factorization.py (hunk 6 above)
Comment (soranjh) [reply pennylane/qchem/factorization.py:186]: Corrected it by replacing unitaries with transformation matrices. The unitaries should be obtained with a new function.

File: tests/qchem/test_factorization.py (hunk 7 above)
Comment (soranjh) [reply tests/qchem/test_factorization.py:174]: The are obtained with the DHF solver, added the code.

File: pennylane/qchem/factorization.py [hunk 8]
```diff
@@ -169,3 +170,152 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+                    )
+                else:
+                    ops_l_ += (
+                        c
+                        * 0.25
+                        * (
+                            qml.Identity(i)
+                            - qml.PauliZ(i)
+                            - qml.PauliZ(j)
+                            + qml.grouping.pauli_mult_with_phase(qml.PauliZ(i), qml.PauliZ(j))[0]
```
Comment (ixfoduap) [inline pennylane/qchem/factorization.py:311]: Curious: why do we need this function from the grouping module? Is this a line that may need to be updated with operator arithmetic?

File: tests/qchem/test_factorization.py [hunk 9]
```diff
@@ -167,3 +167,41 @@ def test_empty_error(two_tensor):
...
+            [
+                np.array([-1.29789639, 0.84064639, 0.45725]),
+                np.array([-4.86900854e-05, -2.75009253e-03, 5.59756523e-03, -2.79878262e-03]),
+                np.array([0.09060524, -0.04530262, -0.04530262]),
+                np.array([-0.34038856, 0.50911518, -0.33745324, 0.16872662]),
+            ],
+        ),
+    ],
+)
+def test_basis_rotation(one_matrix, two_tensor, tol_factor, coeffs_ref):
```
Comment (ixfoduap) [inline tests/qchem/test_factorization.py:203]: I can't think of any more tests at the top of my head, but this seems insufficient for a thorough coverage of this function. I encourage you to spend more time thinking about additional tests we can perform

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: pennylane/qchem/factorization.py [hunk 10]
```diff
@@ -169,3 +170,151 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+
+        .. math::
+
+            H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} T_{pq} a_{p,\alpha}^{\dagger}
+            a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
+            V_{pqrs} a_{p, \alpha}^{\dagger} a_{q, \alpha} a_{r, \beta}^{\dagger} a_{s, \beta},
+
+        where :math:`V_{pqrs}` denotes a two-electron integral in the chemist notation and
+        :math:`T_{pq}` is obtained from the one- and two electron integrals, :math:`h_{pq}` and
+        :math:`h_{pssq}`, as
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py:220]: The original equation is correct. `h_{pssq}` is a component of the two-electron integral that needs to be used to construct the T matrix. This component comes from rearranging the order of fermionic operators to comply with the chemist notation.

File: pennylane/qchem/factorization.py [hunk 11]
```diff
@@ -169,3 +170,151 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+            T_{pq} = h_{pq} - \frac{1}{2} \sum_s h_{pssq}.
+
+        The tensor :math:`V` can be converted to a matrix which is indexed by the indices :math:`pq`
+        and :math:`rs` and eigendecomposed up to a rank :math:`R` to give
+
+        .. math::
+
+            V_{pqrs} = \sum_r^R L_{pq}^{(r)} L_{rs}^{(r) T},
+
+        where :math:`L` denotes the eigenvectors of the matrix. The molecular Hamiltonian can then
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py:233]: Yes, corrected.

File: pennylane/qchem/factorization.py [hunk 12]
```diff
@@ -169,3 +170,151 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+        .. math::
+
+           H = U_0 \left ( \sum_p O_p^{(0)} \right ) U_0^{\dagger} + \sum_r^R U_r \left ( \sum_{q} O_q^{(r)} \right ) U_r^{\dagger},
+
+        where :math:`O = \sum_i c_i P_i` is a linear combination of Pauli words :math:`P_i` that are
+        a tensor product of Pauli :math:`Z` and Identity operators. This allows all the Pauli words
+        in each of the :math:`O` terms to be measured simultaneously. This function returns the
+        coefficients and the Pauli words grouped for each of the :math:`O` terms as well as the
+        eigenvectors of the :math:`T` and :math:`L^{(r)}` matrices that can be used to construct the
+        basis rotation unitaries :math:`U`.
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py:274]: Constructing U merits having dedicated function(s). It might be better to do it in a separate PR?

File: pennylane/qchem/factorization.py [hunk 13]
```diff
@@ -169,3 +170,152 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+                        )
+                    )
+        ops_l.append(ops_l_.tolist())
+
+    ops = [ops_t.tolist()] + ops_l
+
+    c_group = [op.coeffs for op in ops]
+    o_group = [op.ops for op in ops]
+
+    return c_group, o_group, eigvecs
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py:321]: Please see my comment above.

File: pennylane/qchem/factorization.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py]: Corrected.

File: pennylane/qchem/factorization.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/factorization.py]: Done.

File: pennylane/qchem/factorization.py [hunk 14]
```diff
@@ -169,3 +170,147 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+
+    Args:
+        one_electron (array[float]): one-electron integral matrix in the molecular orbital basis
+        two_electron (array[array[float]]): two-electron integral tensor in the molecular orbital
+            basis arranged in chemist notation
+        tol_factor (float): threshold error value for discarding the negligible factors
+
+    Returns:
+        tuple(list[array[float]], list[list[Observable]], list[array[float]]): tuple containing
+        grouped coefficients, grouped ob-servables and basis rotation transformation matrices
```
Comment (ixfoduap) [inline pennylane/qchem/factorization.py:187]:
```suggestion
        grouped coefficients, grouped observables, and basis rotation transformation matrices
```

File: pennylane/qchem/factorization.py [hunk 15]
```diff
@@ -169,3 +170,147 @@ def factorize(two_electron, tol_factor=1.0e-5, tol_eigval=1.0e-5):
...
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [1.398397361, 0.0, 0.0]], requires_grad = False)
+    >>> mol = qml.qchem.Molecule(symbols, geometry)
+    >>> core, one, two = qml.qchem.electron_integrals(mol)()
+    >>> coeffs, ops, eigvecs = basis_rotation(one, two, tol_factor=1.0e-5)
```
Comment (ixfoduap) [inline pennylane/qchem/factorization.py:195]: `eigvecs` should be changed to `unitaries` or something similar, right?

File: tests/qchem/test_factorization.py [hunk 16]
```diff
@@ -167,3 +167,61 @@ def test_empty_error(two_tensor):
 
     with pytest.raises(ValueError, match="All eigenvectors are discarded."):
         qml.qchem.factorize(two_tensor, 1e-5, 1e1)
+
```
Comment (ixfoduap) [inline tests/qchem/test_factorization.py:170]: It's an annoying thing to ask, but I think that this function needs a test that checks that the transformed Hamiltonian is indeed equal to the original one. Namely, that if we take the Hamiltonian expressed in terms of number operators with corresponding coefficients, then apply the transformation unitaries, we recover the original double-factorized Hamiltonian

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Asking for one more annoying-to-implement test

**Review by ixfoduap — APPROVED** [review]

**Review by obliviateandsurrender — APPROVED** [review]
